### PR TITLE
Virtualize history panel

### LINKED ISSUE #1334: History doesnt show anything but in the mouse hover
**Device model, version and app version**

Model Name:   MacBook Pro 16
macOS Version:  Big sur 11.1
App Version: 2021.1 (5)
Safari Version: 14.0.2

**Home Assistant Core Version**

2012.2

**Describe the bug**

App History page shows nothing, except when hovering

**To Reproduce**

**Expected behavior**

Show all
**Screenshots**

App:

Schermafbeelding 2021-01-02 om 23 25 53

App Hovering:

Schermafbeelding 2021-01-02 om 23 26 36

Safari:
Schermafbeelding 2021-01-02 om 23 29 03

**Additional context**

### LINKED ISSUE #3340: History does not render properly with a large number of entities
**Home Assistant release with the issue:** 95.4

**Last working Home Assistant release (if known):**

**Operating environment (Hass.io/Docker/Windows/etc.):**

Docker

**Component/platform:**

**Description of problem:**

When opening the history page,  on my IPad Pro 10.5 (NPDY2LL/A) I get something like:

This happens with iOS Safari or Chrome.  If I click in the white area, I am getting on/off popups for history events.  

If I use any browser in Windows, I’m seeing the expected results. 

**Problem-relevant `configuration.yaml` entries and (fill out even if it seems unimportant):**
```yaml

```

**Traceback (if applicable):**
```

```

**Additional information:**

### LINKED ISSUE #11608: History page doesn’t load
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

My history page does not load. Specifically, it only loads in Chrome on a desktop PC. Will not load on any other browser on the PC and likewise, won’t load on any browser (including HA app) on iPad Pro.

The data loads. And the charts in the second half of the page load on HA if I scroll down, but the bars do not appear and on hover (on a black screen) I can see the pop ups for each sensor, so this is likely a styling/browser issue. Also, when I reduce screen size to 30% width of screen I can see the coloured bars, but the page is unreadable because titles are truncated.

I have had this issue for months, searched high and low and the existing threads I have seen on the issue (4 years old) do not seem to relate to my issue. My history page is heavy, many sensors, switches and automations. Used to be my fave page on HA.

Everything is up-to-date. Running HA on RPi 4 (8G RAM).

Creating issue since others experience same issue on HA community.

### Describe the behavior you expected

Bars do not appear. Seems browser dependent.

### Steps to reproduce the issue

1. Load history page on HA app, or Safari, or Firefox and Chrome on an iPad Pro. Only seems to work on chrome on desktop pc.
2.
3.
...

### What version of Home Assistant Core has the issue?

core-2022.2.3

### What was the last working version of Home Assistant Core?

June 2021

### In which browser are you experiencing the issue with?

All recent

### Which operating system are you using to run this browser?

IPad Pro  OS 15.3

### State of relevant entities

```txt
All
```

### Problem-relevant frontend configuration

```yaml
Source of problem unknown
```

### Javascript errors shown in your browser console/inspector

```txt
None known
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by bdraco**
## Proposed change

Since the history panel didn't load on my instances, and I am doing some more work on the backend history queries, I figure I'd take a stab at this so I can test properly and don't create another regression like https://github.com/home-assistant/core/issues/70346

Also fixes the stalling seen in Chrome when the canvas was still not too large enough for Chrome to reject rendering it.

Fixes #11608 Fixes #3340 Fixes https://github.com/home-assistant/iOS/issues/1334 fixes https://community.home-assistant.io/t/history-page-does-not-load/389576 fixes https://community.home-assistant.io/t/history-graph-is-not-displayed-properly/142775/6 Fixes https://community.home-assistant.io/t/history-does-not-show-on-iphone-12-mini/339863 Fixes https://community.home-assistant.io/t/why-the-heck-dont-the-history-graphs-work-in-the-ios-history-pane-for-1-dimensional-graphs/219632

Screen Shot 2022-05-29 at 13 49 56
Screen Shot 2022-05-29 at 13 49 45

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/chart/state-history-charts.ts [hunk 1]
```diff
@@ -13,6 +13,18 @@ import type { HomeAssistant } from "../../types";
 import "./state-history-chart-line";
 import "./state-history-chart-timeline";
 
+const maxCanvasRows = 256;
```
Comment (bdraco) [inline src/components/chart/state-history-charts.ts:16]: This number might still be too big

File: src/panels/history/ha-panel-history.ts [hunk 2]
```diff
@@ -80,44 +80,44 @@ class HaPanelHistory extends LitElement {
           </app-toolbar>
         </app-header>
 
-        <div class="flex content">
```
Comment (bdraco) [inline src/panels/history/ha-panel-history.ts:83]: Looks like I still have a merge conflict here

Comment (bdraco) [reply src/panels/history/ha-panel-history.ts:83]: never mind. The flex has to be removed for virtualize to work

File: src/components/chart/ha-chart-base.ts [hunk 3]
```diff
@@ -37,6 +37,26 @@ export default class HaChartBase extends LitElement {
...
+  }
+
+  public disconnectedCallback() {
+    this._releaseCanvas();
+    super.disconnectedCallback();
+  }
+
+  public connectedCallback() {
+    super.connectedCallback();
+    if (this.renderRoot && this.renderRoot.querySelector("canvas")) {
```
Comment (bramkragten) [inline src/components/chart/ha-chart-base.ts:55]:
```suggestion
    if (this.hasUpdated) {
```

File: src/components/chart/state-history-charts.ts [hunk 4]
```diff
@@ -48,40 +79,78 @@ class StateHistoryCharts extends LitElement {
...
+      )
+    );
+
+    const combinedItems = Array.prototype.concat.apply(
+      [],
+      [
+        chunkData(this.historyData.timeline, CANVAS_TIMELINE_ROWS_CHUNK),
+        this.historyData.line,
+      ]
+    );
```
Comment (bramkragten) [inline src/components/chart/state-history-charts.ts:100]:
```suggestion
    const combinedItems = chunkData(this.historyData.timeline, CANVAS_TIMELINE_ROWS_CHUNK).concat(
        this.historyData.line
    );
```

File: src/components/chart/state-history-charts.ts [hunk 5]
```diff
@@ -48,40 +79,78 @@ class StateHistoryCharts extends LitElement {
...
   }
 
+  private _renderHistoryItem = (
+    item: TimelineEntity[] | LineChartUnit,
+    index: number
+  ): TemplateResult => {
+    if (!item || index === undefined) {
+      return html``;
+    }
+    if ("unit" in item) {
```
Comment (bramkragten) [inline src/components/chart/state-history-charts.ts:124]: We can just check if it is an array?

File: src/components/chart/state-history-charts.ts [hunk 6]
```diff
@@ -96,18 +165,55 @@ class StateHistoryCharts extends LitElement {
...
+        height: 100%;
+      }
+
       .info {
         text-align: center;
         line-height: 60px;
         color: var(--secondary-text-color);
       }
+      .container {
+        max-height: var(--logbook-max-height);
```
Comment (bramkragten) [inline src/components/chart/state-history-charts.ts:191]:
```suggestion
        max-height: var(--history-max-height);
```

**Comment by bramkragten** [comment]
Nice! Side effect is that we now get a x-axis every 10 rows? Not sure if that is a bad thing perse... Might be nice to have the x-axis scrolling along, but that is def a different PR 😄

File: src/components/chart/state-history-charts.ts (hunk 5 above)
Comment (bdraco) [reply src/components/chart/state-history-charts.ts]: 930aec6a99c8fe2ff752bfa60bc964f30df21c9c

**Comment by bdraco** [comment]
Retesting now

**Comment by bdraco** [comment]
Looks good on local

Testing on 2 prod now

**Comment by bdraco** [comment]
- [ ] m1
- [ ] h1
- [ ] verify repaint after the reconnected

**Comment by bdraco** [comment]
Changes look solid 👍

File: src/components/chart/state-history-charts.ts (hunk 4 above)
Comment (bdraco) [reply src/components/chart/state-history-charts.ts]: Thats way less clunky 👍

**Comment by bdraco** [comment]
Mobile and ipad retest good as well

**Review by bramkragten — DISMISSED** [review]

**Comment by bdraco** [comment]
Retesting firefox now, and will merge if its ok

**Comment by bramkragten** [comment]
Can we add

```
            suggestedMin: this.startTime,
            suggestedMax: this.endTime,
```

To the `x` scale in `state-history-chart-timeline` to make sure all charts have the same start and end, even if they dont have all the data?

**Comment by bdraco** [comment]
Firefox desktop ✅ 

I'm going to recheck safari desktop as well just for completeness

**Comment by bdraco** [comment]
> To the `x` scale in `state-history-chart-timeline` to make sure all charts have the same start and end, even if they dont have all the data?

Testing that now

**Comment by bdraco** [comment]
> Can we add
> 
> ```
>             suggestedMin: this.startTime,
>             suggestedMax: this.endTime,
> ```
> 
> To the `x` scale in `state-history-chart-timeline` to make sure all charts have the same start and end, even if they dont have all the data?
Screen Shot 2022-05-30 at 08 19 25

That causes the bar to be cut off for some reason.

We fetch the history queries with include_start_time_state so we shouldn't have missing data anyways

**Comment by bramkragten** [comment]
We should calculate the max value, as we now use the user selected date, and that is in the future, hence we have no data

**Comment by bdraco** [comment]
I guess that actually makes more sense though since otherwise we are displaying data into the future.

**Comment by bramkragten** [comment]
I also see an issue with the z-index of the mouse over:

CleanShot 2022-05-30 at 20 21 55

**Comment by bdraco** [comment]
> I also see an issue with the z-index of the mouse over:

That one has been a problem for a long time as any canvas the tooltip lines up over has always done that.  Any ideas how to fix it?

**Comment by bramkragten** [comment]
add:
```css
      .entry-container:hover {
        z-index: 1;
      }
```

**Comment by bramkragten** [comment]
I think we have to calculate the suggested endtime like we do for suggested start, based on the data and not on what the user picked.

**Comment by bdraco** [comment]
> I think we have to calculate the suggested endtime like we do for suggested start, based on the data and not on what the user picked.

```
    if (endTime > new Date()) {
      endTime = new Date();
    }
```

that's how we build the data.

I think we can clamp the suggested end time to now

File: src/components/chart/state-history-chart-line.ts [hunk 7]
```diff
@@ -46,6 +46,8 @@ class StateHistoryChartLine extends LitElement {
 
   public willUpdate(changedProps: PropertyValues) {
     if (!this.hasUpdated) {
+      const clampedEndTime =
```
Comment (bramkragten) [inline src/components/chart/state-history-chart-line.ts:49]: Can we do this in `_computedEndTime` in `state-history-chart`?

**Review by bramkragten — DISMISSED** [review]

**Comment by bdraco** [comment]
Retesting now

**Review by bramkragten — DISMISSED** [review]

**Review by bramkragten — DISMISSED** [review]
Hehe missed that 😄 

:tada:

**Comment by bdraco** [comment]
Not sure why its hitting on `Error: src/components/chart/state-history-chart-line.ts(54,13): error TS2322: Type '"time"' is not assignable to type '"timeseries" | undefined'.`

**Comment by bdraco** [comment]
Looks like its not needed for line.  All looks good with the data being clamped

**Comment by bdraco** [comment]
Retest

- [ ] Chrome Mac Desktop
- [ ] Chrome Android
- [ ] iPad (wide)
- [ ] Safari Mac
- [ ] iOS

**Comment by bdraco** [comment]
Retest looks good.  Thanks for the assist @bramkragten  🙌  👍

**Review by bramkragten — DISMISSED** [review]

**Comment by bramkragten** [comment]
`suggestedMax` does work on a line chart, you could change the typing too:

```
  @state() private _chartOptions?: ChartOptions;
```
But am fine with this too 🤷‍♂️

**Comment by bdraco** [comment]
> `suggestedMax` does work on a line chart, you could change the typing too:
> 
> ```ts
>   @state() private _chartOptions?: ChartOptions;
> ```
> 
> But am fine with this too 🤷‍♂️

Works for me.  

Reverted and pushed as b3edbf947afd6147330d435631fd87def4ec9645

**Review by bramkragten — APPROVED** [review]

**Comment by bdraco** [comment]
Retest redux

- [x] Chrome Mac Desktop
- [ ] Firefox
- [ ] Chrome Android
- [ ] iPad (wide)
- [ ] Safari Mac
- [ ] iOS

**Comment by bdraco** [comment]
Everything still looks good.  I'll recheck it again after breakfast

**Comment by bdraco** [comment]
I should test mobile safari outside of the official app as well

- [ ] mobile safari
- [ ] mobile brave
- [ ] mobile Firefox

**Comment by bdraco** [comment]
Looks good.  There is a 1-s delay with large datasets as the chart data is constructed after the data is received from the backend where the cpu is swamped but I don't think we can do anything about that since we will also have to mutate the data into a format we can render.  We can probably optimize the data mutation in the future but this is an existing issue.  It's already way faster than 2022.5 since we used the compressed history data format from the websocket.  Something for the future

**Pull request merged**